planner/core/rule_partition_processor.go

```diff
@@ -1575,13 +1576,23 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
 		return 0, len(p.data), false
 	}
 
-	start, end := p.pruneUseBinarySearch(sctx, opName, con, op)
+	// If different collation, we can only prune if:
+	// - expression is binary collation (can only be found in one partition)
+	// - EQ operator, consider values 'a','b','ä' where 'ä' would be in the same partition as 'a' if general_ci, but is binary after 'b'
+	// otherwise return all partitions / no pruning
+	_, exprColl := expr.CharsetAndCollation()
+	colColl := p.partCol.RetType.Collate
+	if exprColl != "" && exprColl != colColl && (opName != ast.EQ || !collate.IsBinCollation(exprColl)) {
```

Comment:
I think we can remove the condition `exprColl != ""`, even if the `exprColl` could be `""`, it should be 
```suggestion
	if exprColl == "" || exprColl != colColl && (opName != ast.EQ || !collate.IsBinCollation(exprColl)) {
```